### PR TITLE
schemachanger,backupccl: support for backup and restore mid-change

### LINKED ISSUE #73071: sql/schemachanger,backupccl: plot a path for adoption of the declarative schema changer
**Overview**

Restores currently jump through some hoops to leverage the existing, imperative schema changer when restoring descriptors which are currently mid-schema change. These same mechanisms won't perfectly work for the declarative schema changer. This issue encompasses both
* plotting a plan for restoring descriptors which are mid-schema change in the declarative schema changer, and 
* restoring descriptors which preceded the declarative schema changer after we've removed the old code.

##### Background 

In the old world, schema changes operated on a per-descriptor basis and all of their state was implied by the descriptor itself (modulo some description sort of things in the job). This was something of a boon to BACKUP/RESTORE; from a descriptor, the job could be recreated relatively trivially as it didn’t have much state (not that trivially, there’s still a bunch of complexity related to filtering out things that maybe don’t exist). 

The above scheme has several known problems which we're trying to fix:
* Schema changes need to be grouped by transactions and we’d like cross-descriptor rollback or completion.
* The tight coupling of implicit schema changer state and descriptor logical/physical state has led to much confusion and bad code.
* The queuing of mutations is devilishly complex.

The declarative schema changer seeks to move away from the descriptor mutation world. Changes will be decomposed into sets of schema elements to be added or removed. These elements will traverse statuses. The planner will formulate a plan based solely on the current set of (element, direction, status) tuples. This is nice because it lets us decompose a bunch of logic in order to build more robust primitives. 

A downside is that, from the descriptor alone, we can’t really reconstruct the job state at full fidelity. For a little while, we probably can reconstruct it sufficiently. At the end of the day, we’re not eager to break compatibility with existing descriptors (way too much headache for their users) and we’re trying to not stray too far from the concepts of descriptors which already exist.

##### Problem statement

At some point down the road, we’ll want to restore a descriptor which preceded the declarative schema changer, yet, we’ll also want to have deleted the schema changer code as it existed when that descriptor was created.
At a closer point, we’ll want to restore descriptors which are undergoing schema changes using the declarative schema changer.

**Describe the solution you'd like**

One family of solutions involves moving more of the schema change state back onto the descriptors. One could imagine that we retain an in-sync set of elements and their statuses on the descriptors and in the job. One could go further and just have the job reference the descriptors. This actually wouldn't be that big of a change. There'd still be some work to figure out which elements to trash when restoring. This would be new work, but it wouldn't be too crazy, I don't think. That would give us a path forward in terms of restoring these descriptors in the absence of jobs.

Separately, we could then write migrations which synthesize elements and their statuses from mutations. This does not need to be done until we want to remove the old schema changer.

**Additional context**

* By having all of the state in the descriptors instead of in the jobs, we may run afoul of other versioning considerations for full cluster restore.
    * Need to really nail down the model for jobs crossing versions.
 
* In the transactional schema change RFC, there’s a notion of a descriptor which is in a state that needs to be rolled back. This is a bit hand-wavy in the RFC.
    * There are tradeoffs here regarding how exactly to atomically add constraints without exposing them to the client before the constraint is committed.
    * In general, there’s a feeling that it’d be okay to ship a version of transactional schema changes which exposed, very briefly, clients to enforcing a constraint which has not yet been committed. The alternative, more complex, approach is to have the execution of the statement preempt the schema change. This would probably be better, but not definitely.
         * “There’s isolation, then there’s isolation”
    * This cleanup differs from what we’d want to do in restore in that it always assumes we want to rollback whereas restore may want to roll forward.

### PR DESCRIPTION
**Pull request by ajwerner**
### schemachanger,backupccl: support for backup and restore mid-change

This commit touches the schema changer to encode some more metadata about the
state of the schema change into the descriptors in order to better facilitate
synthesizing a job at restore time.

The major contribution, then, is to support the declarative schema changer in
restore. Along the way, it picks up some catalog niceties like batches
descriptor retreival and the nstree.Catalog.

Finally, the code is tested with a suite of datadriven tests which run schema
changes and then take the state and BACKUP/RESTORE it at each of the possible
stages (including all of the rollback stages) and makes sure that the right
outcome ultimately happens.

Note that this is currently just tested for database-level BACKUP/RESTORE
and that more testing is planned as follow-up.

### sql/schemachanger: move where we remove namespace entries to be earlier

This should really use the namespace element, but I needed this fix
to not break some other tests I'm writing.

### backupccl: don't back up dropped databases or schemas

Hack to deal with bugs that Aditya is fixing concurrently in https://github.com/cockroachdb/cockroach/pull/76635.

### sql/schemachanger: add more end-to-end testing

This reworking is a precursor to adding backup/restore testing. Pause testing
is also extremely valuable. It caught a bug when we resume after having
finished a backfill but before we marked it as complete. This was bad
for two reasons:

1) We weren't catching the panic on those goroutines (lesson learned, again)
2) We weren't detecting the case where we were done.

Fixes #73071 

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
@dt would you accept a method on the `ctxgroup` which takes an `op` string and does a `defer ... recover` to catch panics?

**Comment by dt** [comment]
> @dt would you accept a method on the ctxgroup which takes an op string and does a defer ... recover to catch panics?

Can you expand on motivation? Don't we generally eschew `recover`?

**Comment by ajwerner** [comment]
> > @dt would you accept a method on the ctxgroup which takes an op string and does a defer ... recover to catch panics?
> 
> Can you expand on motivation? Don't we generally eschew `recover`?

Do we?  https://github.com/cockroachdb/cockroach/issues/58164 / https://github.com/cockroachdb/cockroach/pull/62243 /  it seems better to catch them than not and crash the server. These days that's quite the taboo.

In the first commit, it turned out that if you pause and resume a backfill in the new schema changer after you've checkpointed the progress of completing everything but before marking the state change, then we'd resume and construct an empty set of spans to do, which hits the below panic (should that really be a panic?). Anyway, it'd be fine if we caught that but is rather bad that it crashes the server. I added the below snippet to the code to prevent such things in the future, but it's a pretty heavyweight pattern to expect folks to remember every time:

```go
			defer func() {
				switch r := recover().(type) {
				case nil:
					return
				case error:
					err = errors.Wrapf(r, "failed to %s", op)
				default:
					err = errors.AssertionFailedf("failed to %s: %v", op, r)
				}
			}()
```

https://github.com/cockroachdb/cockroach/blob/8ba916b2e4c6c20977baf115dbbba43957ccae0f/pkg/sql/distsql_physical_planner.go#L960

**Comment by ajwerner** [comment]
> This doesn't feel like it belongs in this change. If this change was hitting panics, it should just check len(spans) != 0 before calling partitionSpans and avoid the panic, rather than hit and recover it, right?

I *also* did that. Yes, that's definitely a right thing to do, but it seems stupid to not defensively translate the panics, as we do during all query execution and job execution.

**Comment by ajwerner** [comment]
> If you want to make a general argument that we're using ctxgroup all over bulk ops and schema changes and that across all these we don't want operations in the workers that panic to kill the whole proc, we should make that argument in a standalone issue or change, right?

Yes, sorry for using this PR as a conduit for a conversation about a separate change, it was just top-of-mind.

**Comment by ajwerner** [comment]
https://github.com/cockroachdb/cockroach/issues/76734 is the issue.

**Review by fqazi — CHANGES_REQUESTED** [review]
This looks really good, just very minor things that can be dealt with later

> Reviewed 7 of 7 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, 18 of 22 files at r4, 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-:-Mw82nRQ8qnh-MxzuFHE:b-ky39dx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @dt)

___
*[pkg/ccl/backupccl/restore_planning.go, line 1231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw7wr9hEDNPbIXXyAIc:-Mw7wr9hEDNPbIXXyAId:b-yvuim6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/ccl/backupccl/restore_planning.go#L1231)):*
> ```Go
> 		}
> 		var ids []descpb.ID
> 		switch e := t.Element().(type) {
> ```

This stuff does worry me, currently, I think it's okay. But longer-term it feels like something will be missed and cause breakages. Maybe some how have attributes for these in the future
___
*[pkg/sql/schemachanger/sctest/cumulative.go, line 454 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw82-4aBsx76FvHZnCU:-Mw82-4aBsx76FvHZnCV:b-ouz0kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/sql/schemachanger/sctest/cumulative.go#L454)):*
> ```Go
>              WHERE "parentID" = (SELECT id FROM system.namespace WHERE name = $1)
>         )`, dbName)
> 			afterRestore := tdb.QueryStr(t, fetchDescriptorStateQuery)
> ```

I guess this isn't an issue right now, but any danger from fields that need to be filtered? I guess we would need to regex them out (things like timestamps)

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @postamar)

___
*[pkg/ccl/backupccl/restore_planning.go, line 1231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw7wr9hEDNPbIXXyAIc:-Mw86is5CqAAU7JCTk7Y:bct2w4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/ccl/backupccl/restore_planning.go#L1231)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This stuff does worry me, currently, I think it's okay. But longer-term it feels like something will be missed and cause breakages. Maybe some how have attributes for these in the future
</blockquote></details>

Me too, I'll put a TODO. @postamar reports that these slices are going away.
___
*[pkg/sql/schemachanger/sctest/cumulative.go, line 454 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw82-4aBsx76FvHZnCU:-Mw86oOs7dkMIbUUcakz:b-3a08qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/sql/schemachanger/sctest/cumulative.go#L454)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I guess this isn't an issue right now, but any danger from fields that need to be filtered? I guess we would need to regex them out (things like timestamps)
</blockquote></details>

this is using the `SHOW CREATE` strings to compare the tables as opposed to the descriptor values for exactly that reason. It's not perfect, but it's a pretty good proxy for the actual state of the descriptor. 

I'm going to add a call to also validate all of the descriptors in the cluster here and call it a day.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @postamar)

___
*[pkg/ccl/backupccl/restore_planning.go, line 1231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw7wr9hEDNPbIXXyAIc:-Mw8CA104oUTVZ5vftsG:b-3pexiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/ccl/backupccl/restore_planning.go#L1231)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Me too, I'll put a TODO. @postamar reports that these slices are going away.
</blockquote></details>

I'm realizing that this needs to find all of the expressions in the structs too and it needs to rewrite those. Maybe reflection is in order.

File: pkg/sql/schemachanger/sctest/cumulative.go [hunk 1]
```diff
@@ -23,24 +23,31 @@ import (
...
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scrun"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/datadriven"
-	"github.com/pkg/errors"
+	"github.com/cockroachdb/errors"
+	"github.com/lib/pq"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
```
Comment (dt) [inline pkg/sql/schemachanger/sctest/cumulative.go:39]: 2¢ I just use ctxgroup even if I don't need the ctx yet.

File: pkg/ccl/backupccl/restore_planning.go [hunk 2]
```diff
@@ -1170,6 +1185,62 @@ func rewriteSchemaDescs(schemas []*schemadesc.Mutable, descriptorRewrites DescRe
...
+		if err := rewriteSchemaChangerState(
+			sc.DeclarativeSchemaChangerState, descriptorRewrites,
+		); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func rewriteSchemaChangerState(
```
Comment (dt) [inline pkg/ccl/backupccl/restore_planning.go:1198]: I'm tempted to say rewriteSchemaChangerState should be in an sc package. 


I mean, I'm even tempted to say all these existing rewrite methods that dig into catalog elements and replace one ID with another should move to an sc package that restore just calls into, but I understand that might be out of scope here, so I guess this adding one here along side keeps it consistent.

File: pkg/ccl/backupccl/restore_job.go [hunk 3]
```diff
@@ -2011,6 +1997,106 @@ func (r *restoreResumer) publishDescriptors(
...
+				"version mismatch for descriptor %d, expected version %d, got %v",
+				got[i].GetID(), got[i].GetVersion(), expVersion[id],
+			)
+		}
+		all.UpsertDescriptorEntry(got[i])
+	}
+	return all.Catalog, nil
+}
+
+// createDeclarativeSchemaChangeJobs is called during the last phase of a
```
Comment (dt) [inline pkg/ccl/backupccl/restore_job.go:2051]: this seems like a good candidate to be in a sc package. could genericize the comment slightly if it was moved to such, to say "when ingesting external descriptors such as in the last phase of restore"

**Review by dt — COMMENTED** [review]
only looked at 4th commit and I only looked closely-ish at the backupccl bits but overall I like it, and really like the move towards calling specific helpers over just tons and tons of inline descriptor work in big restore functions.

File: pkg/ccl/backupccl/restore_planning.go (hunk 2 above)
Comment (ajwerner) [reply pkg/ccl/backupccl/restore_planning.go:1198]: Sure

Comment (ajwerner) [reply pkg/ccl/backupccl/restore_planning.go:1198]: Well, I thought about it and a lot of symbols would have to move. I directionally agree, but not for this PR.

File: pkg/ccl/backupccl/restore_job.go (hunk 3 above)
Comment (ajwerner) [reply pkg/ccl/backupccl/restore_job.go:2051]: I moved this out to a new package `scbackup`.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @fqazi, and @postamar)

___
*[pkg/ccl/backupccl/restore_planning.go, line 1231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-Mw7wr9hEDNPbIXXyAIc:-Mw9y6rYC7zkEUy7tyQ4:boaumjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec814da1f4e9c2c1d8b6e21f98ad96744ce871d/pkg/ccl/backupccl/restore_planning.go#L1231)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm realizing that this needs to find all of the expressions in the structs too and it needs to rewrite those. Maybe reflection is in order. 
</blockquote></details>

Okay, the reflection dust has been sprinkled onto this.

**Review by fqazi — APPROVED** [review]
Took a quick look at the latest set of commits. Happy with the reflection-based approach! :lgtm: 

> Reviewed 26 of 26 files at r7, 5 of 5 files at r8, 13 of 13 files at r9, 3 of 3 files at r10, 24 of 24 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-:-MwA1ve27PtcMw9qSE7F:b5vvmzg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/screl/walk.go, line 24 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-MwA0Oww66nY3iIzXcGz:-MwA0Oww66nY3iIzXcH-:b-uth4yp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8fa8c1d8fb5dbb86b9ce966d98ae13fdaa2e025/pkg/sql/schemachanger/screl/walk.go#L24)):*
> ```Go
> 
> // WalkDescIDs calls f for every catid.DescID field in e.
> func WalkDescIDs(e scpb.Element, f func(id *catid.DescID) error) error {
> ```

Love this approach much cleaner!
___
*[pkg/sql/schemachanger/screl/walk_test.go, line 31 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76715#-MwA0XMA4bKcnfgYObUh:-MwA0XMA4bKcnfgYObUi:b5vp05a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8fa8c1d8fb5dbb86b9ce966d98ae13fdaa2e025/pkg/sql/schemachanger/screl/walk_test.go#L31)):*
> ```Go
> // as advertised.
> func TestWalk(t *testing.T) {
> ```

Nit: extra new line

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**